sklearn/neighbors/lmnn.py

```diff
@@ -0,0 +1,1013 @@
+                     '`n_neighbors_` will be set to {} for estimation.'
+                     .format(self.n_neighbors, min_non_singleton_size,
+                             min_non_singleton_size-1))
+
+            self.n_neighbors_ = min(self.n_neighbors, min_non_singleton_size-1)
+        else:
+            self.n_neighbors_ = targets.shape[1]
+
+        return X, y_inverse, targets, init
+
+    @staticmethod
+    def _compute_grad_static(X, targets, verbose=0):
+        """Compute the gradient component due to the target neighbors that
+        stays fixed throughout training
+
+        Parameters
+        ----------
+        X : array, shape (n_samples, n_features)
+            The training samples.
+
+        targets : array, shape (n_samples, n_neighbors)
+            The k nearest neighbors of each sample from the same class.
+
+        verbose : int, optional (default=0)
+            If not 0, progress info will be printed.
+
+        Returns
+        -------
+        grad_static, shape (n_features, n_features)
+            An array with the sum of all weighted outer products.
+        """
+        if verbose:
+            print('Computing static part of the gradient...')
+
+        n_samples, n_neighbors = targets.shape
+        row = np.repeat(range(n_samples), n_neighbors)
+        col = targets.ravel()
+        targets_sparse = csr_matrix((np.ones(targets.size), (row, col)),
+                                    shape=(n_samples, n_samples))
+
+        return sum_outer_products(X, targets_sparse)
+
+    def _lbfgs_callback(self, transformation):
+        if self.callback is not None:
+            self.callback(transformation, self.n_iter_)
+
+        self.n_iter_ += 1
+
+    def _loss_grad(self, transformation, X, y, targets, grad_static):
+        """Compute the loss under a given ``transformation`` and the
+        loss gradient w.r.t. ``transformation``.
+
+        Parameters
+        ----------
+        transformation : array, shape (n_features_out * n_features,)
+            The current (flattened) linear transformation.
+
+        X : array-like, shape (n_samples, n_features)
+            The training samples.
+
+        y : array, shape (n_samples,)
+            The corresponding training labels.
+
+        targets : array, shape (n_samples, n_neighbors)
+            The target neighbors of each sample.
+
+        grad_static : array, shape (n_features, n_features)
+            The gradient component caused by target neighbors, that stays
+            fixed throughout the algorithm.
+
+        Returns
+        -------
+        loss: float
+            The new loss.
+        grad: array, shape (n_features_out * n_features,)
+            The new (flattened) gradient.
+        """
+
+        n_samples, n_features = X.shape
+        self.transformation_ = transformation.reshape(-1, n_features)
+
+        t_start = time.time()
+        X_embedded = self.transform(X, check_input=False)
```

Comment:
we don't expose check_input in transform method. I would use a private _transform method that has check_input and stick to transform(X)